File: sorting-searching/sorting-algorithms/burst-sort/burst-sort.py
BURST_THRESHOLD = 3

def burstsort_recursive(strings, depth, result):
    if not strings:
        return

    if len(strings) <= BURST_THRESHOLD:
        result.extend(sorted(strings))
        return

    buckets = {}

    for string in strings:
        key = string[depth] if depth < len(string) else "\0"

        if key not in buckets:
            buckets[key] = []

        buckets[key].append(string)

    for key in sorted(buckets.keys()):
        burstsort_recursive(
            buckets[key],
            depth + 1,
            result
        )


def burstsort(strings):
    result = []

    burstsort_recursive(strings, 0, result)

    return result


words = [
    "apple",
    "apply",
    "ape",
    "april",
    "banana",
    "band",
    "ball"
]

print("Original strings:")
print(words)

sorted_words = burstsort(words)

print("\nSorted strings:")
print(sorted_words)

File: sorting-searching/sorting-algorithms/burst-sort/burst-sort.js
const BURST_THRESHOLD = 3;

function burstsortRecursive(strings, depth, result) {
  if (strings.length === 0) {
    return;
  }

  if (strings.length <= BURST_THRESHOLD) {
    const sorted = [...strings].sort();

    result.push(...sorted);

    return;
  }

  const buckets = new Map();

  for (const str of strings) {
    const key = depth < str.length ? str[depth] : "\0";

    if (!buckets.has(key)) {
      buckets.set(key, []);
    }

    buckets.get(key).push(str);
  }

  const keys = [...buckets.keys()].sort();

  for (const key of keys) {
    burstsortRecursive(
      buckets.get(key),
      depth + 1,
      result
    );
  }
}

function burstsort(strings) {
  const result = [];

  burstsortRecursive(strings, 0, result);

  return result;
}

const words = [
  "apple",
  "apply",
  "ape",
  "april",
  "banana",
  "band",
  "ball"
];

console.log("Original strings:");
console.log(words);

const sortedWords = burstsort(words);

console.log("\nSorted strings:");
console.log(sortedWords);
